src/wh_server_dma.c

```diff
@@ -0,0 +1,307 @@
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_error.h"
+#include "wolfhsm/wh_server_dma.h"
+
+/* TODO: if the Address allowlist ever gets large, we should consider a more
+ * efficient representation (requiring sorted array and binary search, or
+ * building a binary tree, etc.) */
+
+
+static int _checkAddrAgainstAllowList(const whServerDmaAddrList allowList, void* addr,
+                                      size_t size)
+{
+    uintptr_t startAddr = (uintptr_t)addr;
+    uintptr_t endAddr   = startAddr + size;
+
+    /* Check if the address range is fully within a allowlist entry */
+    for (int i = 0; i < WH_DMA_ADDR_ALLOWLIST_SIZE; i++) {
+        uintptr_t allowlistStartAddr = (uintptr_t)allowList[i].addr;
+        uintptr_t allowlistEndAddr   = allowlistStartAddr + allowList[i].size;
+
+        if (startAddr >= allowlistStartAddr && endAddr <= allowlistEndAddr) {
+            return WH_ERROR_OK;
+        }
+    }
+
+    return WH_ERROR_ACCESS;
+}
+
+static int _checkMemOperAgainstAllowList(const whServerDmaAddrAllowList* allowList,
+                                         whServerDmaOper oper, void* addr,
+                                         size_t size)
+{
+    int rc = WH_ERROR_OK;
+
+    /* If a read/write operation is requested, check the transformed address
+     * against the appropriate allowlist
+     *
+     * TODO: do we need to allowlist check on POST in case there are subsequent
+     * memory operations for some reason?
+     */
+    if (oper == WH_DMA_OPER_CLIENT_READ_PRE) {
```

Comment:
Consider switch() with default to return WH_ERROR_BADARGS if the oper is unknown.

Reply:
Right now the switch would have empty cases for READ_POST and WRITE_POST, as we don't do anything with those. Instead, I added `isOperValid()` as a check to this function and return failure if an invalid operation is requested. LMK what you think